<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Simon Klau" />

<meta name="date" content="2023-03-30" />

<title>prioritylasso Vignette</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">prioritylasso Vignette</h1>
<h4 class="author">Simon Klau</h4>
<h4 class="date">2023-03-30</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>prioritylasso was developed for situations in which different types
of high dimensional omics-data in combination with clinical data are
available. We want to include these so called blocks of data together in
a prediction model. prioritylasso thereby tries to find a good
compromise between prediction accuracy and practical aspects. This is
done by letting the practitioner include prior knowledge in the form of
priorities, which define the importance of every block. The priorities
can be chosen in several different ways.<br />
The model is successively calculated as LASSO-based, while the Linear
Predictor after each fit is taken as an Offset in the LASSO Regression
of the block with the next lowest priority. The Penalized Regression
models are computed via the R package <code>glmnet</code>. Moreover, the
R package <code>survival</code> is used when the outcome consists of
survival data. In addition to the main function, the package
<code>prioritylasso</code> contains an extension of the standard
function named cvm_prioritylasso. It is also explained under practical
aspects in the following examples.</p>
</div>
<div id="getting-started-with-prioritylasso" class="section level2">
<h2>Getting started with prioritylasso</h2>
<p>The data we will use for the examples was simulated and shall
represent AML data with 4 blocks of variables. We first have to load the
package which contains this data set.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(prioritylasso)</span></code></pre></div>
<p>To take a deeper look at the data, open the description through
<code>?pl_data</code>. The different types of data are all stored
together. To get and define the block structure which implies the
priorities, we look at the names of the variables and the matrix
dimension.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">dim</span>(pl_data)</span></code></pre></div>
<pre><code>## [1]  400 1029</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">colnames</span>(pl_data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;b1.1&quot;  &quot;b1.2&quot;  &quot;b1.3&quot;  &quot;b1.4&quot;  &quot;b2.1&quot;  &quot;b2.2&quot;  &quot;b2.3&quot;  &quot;b2.4&quot;  &quot;b2.5&quot; 
## [10] &quot;b3.1&quot;  &quot;b3.2&quot;  &quot;b3.3&quot;  &quot;b3.4&quot;  &quot;b3.5&quot;  &quot;b3.6&quot;  &quot;b3.7&quot;  &quot;b3.8&quot;  &quot;b3.9&quot; 
## [19] &quot;b3.10&quot; &quot;b3.11&quot; &quot;b3.12&quot; &quot;b3.13&quot; &quot;b3.14&quot; &quot;b3.15&quot; &quot;b3.16&quot; &quot;b3.17&quot; &quot;b3.18&quot;
## [28] &quot;b3.19&quot; &quot;b4.1&quot;  &quot;b4.2&quot;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">colnames</span>(pl_data)[<span class="dv">1025</span><span class="sc">:</span><span class="dv">1029</span>]</span></code></pre></div>
<pre><code>## [1] &quot;b4.997&quot;  &quot;b4.998&quot;  &quot;b4.999&quot;  &quot;b4.1000&quot; &quot;pl_out&quot;</code></pre>
<p>We see that the last column contains the outcome. We will re-save the
outcome and the matrix of predictors separately and define the argument
<code>blocks</code> afterwards. We will need that for the application of
prioritylasso where the indices have to correspond with the variables in
the predictor matrix.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>pl_out <span class="ot">&lt;-</span> pl_data[,<span class="dv">1029</span>]</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>pl_pred <span class="ot">&lt;-</span> pl_data[,<span class="dv">1</span><span class="sc">:</span><span class="dv">1028</span>]</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>blocks <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">bp1=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">bp2=</span><span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">bp3=</span><span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="at">bp4=</span><span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>)</span></code></pre></div>
<p>For an easier interpretation, we can consider the first block as
clinical variables of most importance, the second block as other
clinical data, and the third block as mutations. These blocks are all
low-dimensional. That will later be an advantage regarding computation
time. Another type of data is high dimensional gene expression data.
Moreover, <code>pl_out</code> consists of a binary outcome for 400
subjects. Before we go further we want to split the data into training
and validation set. The training set should be used to build the model
with prioritylasso while the validation data should later be used to
assess the prediction accuracy in an independent data set. Here we just
do a random split such that 2/3 of the data belongs to the training set
and 1/3 to the validation set.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>label <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">dim</span>(pl_pred)[<span class="dv">1</span>],<span class="fu">round</span>(<span class="fu">dim</span>(pl_pred)[<span class="dv">1</span>]<span class="sc">*</span>(<span class="dv">2</span><span class="sc">/</span><span class="dv">3</span>)))</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>pl_train <span class="ot">&lt;-</span> pl_pred[label,]</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>pl_val <span class="ot">&lt;-</span> pl_pred[<span class="sc">-</span>label,]</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>pl_out_train <span class="ot">&lt;-</span> pl_out[label]</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>pl_out_val <span class="ot">&lt;-</span> pl_out[<span class="sc">-</span>label]</span></code></pre></div>
<p>We can now run prioritylasso for the first time.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>pl1 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, </span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>                     <span class="at">blocks =</span> blocks, <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Because we have a binary outcome, we specified
<code>family = &quot;binomial&quot;</code> and <code>type.measure = &quot;auc&quot;</code>.
The priority structure was given through the block definition in the
data set. An extended use of the block definition and the general
functionality of prioritylasso will be given in the following
sections.</p>
</div>
<div id="prioritylasso---extended-application" class="section level2">
<h2>prioritylasso - extended application</h2>
<div id="specifying-the-priorities" class="section level4">
<h4>Specifying the priorities</h4>
<p>If we want to specify another block order, we have to redefine the
argument <code>blocks</code>. Let us assume that we do not want to
include the variables b1 first in the model, but b3 first, b1 second and
b2 third. That would be done by
<code>blocks = list(bp1 = 10:28, bp2 = 1:4, bp3 = 5:9, bp4 = 29:1028)</code>.
So the position on the list indicates when the variables are considered
in the model or in other words, the priority. It is not necessary to
name the entries of the list, but it may help in avoiding confusion. We
chose the names such that “bp1”” means “block of priority 1”.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>pl2 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, </span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                     <span class="at">block1.penalization =</span> <span class="cn">TRUE</span>, <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>,  <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>),</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                     <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>pl2<span class="sc">$</span>nzero</span></code></pre></div>
<pre><code>## [[1]]
## [1] 17
## 
## [[2]]
## [1] 4
## 
## [[3]]
## [1] 2
## 
## [[4]]
## [1] 23</code></pre>
<p>The output lists are to be interpreted according to the definition in
<code>blocks</code>. Therefore, the first entry corresponds to the
variables of columns 10:28 in the data set.</p>
</div>
<div id="set-a-maximal-number-of-nonzero-coefficients." class="section level4">
<h4>Set a maximal number of nonzero coefficients.</h4>
<p>As we can see in pl2, we have a lot of nonzero coefficients of bp4,
although it is the block with lowest priority. In general, there should
be a reason why the glmnet procedure chooses these coefficients. On the
other hand, it is sometimes not appropriate from a practical point of
view that the number exceeds a particular value. Therefore we made it
possible to set a maximal number of nonzero coefficients. Because we
just want to set a limit for bp4, we set the other entries to
<code>Inf</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>pl3 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, </span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                     <span class="at">block1.penalization =</span> <span class="cn">TRUE</span>, <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>,  <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>                     <span class="at">max.coef =</span> <span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>), <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>pl3<span class="sc">$</span>nzero</span></code></pre></div>
<pre><code>## [[1]]
## [1] 17
## 
## [[2]]
## [1] 4
## 
## [[3]]
## [1] 2
## 
## [[4]]
## [1] 10</code></pre>
</div>
<div id="include-the-block-with-the-highest-priority-without-a-penalty" class="section level4">
<h4>Include the block with the highest priority without a penalty</h4>
<p>We can further diversify the option of whether or not bp1 is included
without a penalty in the model. In a not penalized version, the block
has to be low-dimensional and the model includes all of its variables.
In this case, the model fit is stored in <code>block1unpen</code>.
Otherwise, the block is treated like the other blocks and a LASSO model
is fitted.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>pl4 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, </span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>                     <span class="at">block1.penalization =</span> <span class="cn">FALSE</span>, <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>                     <span class="at">max.coef =</span> <span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>), <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>pl4<span class="sc">$</span>block1unpen</span></code></pre></div>
<pre><code>## 
## Call:  glm(formula = Y[current_observations] ~ X[current_observations, 
##     blocks[[1]]], family = family, weights = weights[current_observations])
## 
## Coefficients:
##           b1.1    b1.2    b1.3    b1.4  
## 0.1683  1.0135  0.9081  1.0682  0.3760  
## 
## Degrees of Freedom: 266 Total (i.e. Null);  262 Residual
## Null Deviance:       351 
## Residual Deviance: 336.8     AIC: 346.8</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>pl4<span class="sc">$</span>lambda.ind</span></code></pre></div>
<pre><code>## [[1]]
## NULL
## 
## [[2]]
## [1] 5
## 
## [[3]]
## [1] 10
## 
## [[4]]
## [1] 14</code></pre>
<p>The first entries of the values which correspond to the
<code>blocks</code> list remain empty as you can see for the example of
<code>lambda.ind</code>.</p>
</div>
<div id="options-for-the-cross-validation-procedure" class="section level4">
<h4>Options for the cross validation procedure</h4>
<p><code>nfolds</code> and <code>lambda.type</code> are options which
refer to the cross-validation procedure in <code>cv.glmnet</code>.
<code>nfolds</code> specifies the number of folds in the procedure while
<code>lambda.type</code> handles the amount of cross-validation error.
It can be set to either <code>lambda.min</code>, which is also the
default, or <code>lambda.1se</code>. <code>lambda.min</code> gives the
result with minimum mean cross-validation error, whereas
<code>lambda.1se</code> gives the result such that the cross-validation
error is within 1 standard error of the minimum, and thus leads to more
sparse results. Note that the latter can only be chosen in combination
with no restrictions in <code>max.coef</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>pl5_min <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>                         <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, <span class="at">block1.penalization =</span> <span class="cn">TRUE</span>, </span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>                         <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>                         <span class="at">lambda.type =</span> <span class="st">&quot;lambda.min&quot;</span>, <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>pl5_1se <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>                         <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, <span class="at">block1.penalization =</span> <span class="cn">TRUE</span>, </span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>                         <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>                         <span class="at">lambda.type =</span> <span class="st">&quot;lambda.1se&quot;</span>, <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a>pl5_min<span class="sc">$</span>nzero</span></code></pre></div>
<pre><code>## [[1]]
## [1] 4
## 
## [[2]]
## [1] 1
## 
## [[3]]
## [1] 5
## 
## [[4]]
## [1] 55</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>pl5_1se<span class="sc">$</span>nzero</span></code></pre></div>
<pre><code>## [[1]]
## [1] 3
## 
## [[2]]
## [1] 0
## 
## [[3]]
## [1] 0
## 
## [[4]]
## [1] 16</code></pre>
<p>In addition, other options for prioritylasso can be specified. For
example, we can change the type of cross validation measurement for
binary outcome to the misclassification error. Examples for other
outcome variables are shown in the function documentation. Further
arguments can be passed to prioritylasso which are used in
<code>cv.glmnet</code>, e.g. the elasticnet mixing parameter. Here, we
are only using the default value 1 which corresponds to the standard
LASSO method, but a parameter <code>alpha</code> between 0 and 1 can be
chosen as well.</p>
</div>
<div id="some-notes-on-the-output" class="section level4">
<h4>Some notes on the output</h4>
<p>The function used for every LASSO fit is glmnet which creates a
sequence of lambda values. The lambda which is chosen according to
<code>lambda.type</code> is the lambda on position
<code>lambda.ind</code> of the sequence and its value is stored in
<code>lambda.min</code>. In general, the lower the value of
<code>lambda.ind</code>, the higher the <code>lambda.min</code> and thus
the penalization. This leads to more sparse models. The number of
<code>lambda</code> values can be chosen with an optional argument. We
give the output of pl1 as an example.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>pl1<span class="sc">$</span>lambda.ind</span></code></pre></div>
<pre><code>## [[1]]
## [1] 26
## 
## [[2]]
## [1] 5
## 
## [[3]]
## [1] 8
## 
## [[4]]
## [1] 49</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>pl1<span class="sc">$</span>lambda.min</span></code></pre></div>
<pre><code>## [[1]]
## [1] 0.002421659
## 
## [[2]]
## [1] 0.05201088
## 
## [[3]]
## [1] 0.0129989
## 
## [[4]]
## [1] 0.0213665</code></pre>
<p>The values of lambda are generated in every call of glmnet and hence
cannot be compared. The corresponding mean cross validated error is
stored in the list <code>min.cvm</code>. The interpretation of its
values depend on the argument <code>type.measure</code>. In our examples
with a binary outcome it is usually the area under the ROC curve
(AUC).</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>pl1<span class="sc">$</span>min.cvm</span></code></pre></div>
<pre><code>## [[1]]
## [1] 0.6197587
## 
## [[2]]
## [1] 0.638546
## 
## [[3]]
## [1] 0.6735664
## 
## [[4]]
## [1] 0.8342153</code></pre>
<p>We see that for the example of pl1 it grows with the number of
considered modalities. That is what we expected because the more
coefficients we have in our model, the better the prediction should
be.</p>
</div>
</div>
<div id="use-of-the-function-cvm_prioritylasso" class="section level2">
<h2>Use of the function cvm_prioritylasso</h2>
<p>prioritylasso was generally implemented with the idea, that
practitioners have prior knowledge about the data that allows them to
specify the priorities. However, it might be, especially in the presence
of several blocks, that it is not clear which order of blocks is the
best. That is why we implemented cvm_prioritylasso that allows us to
define more than one possible list of blocks. The function chooses the
best block order according to the mean cross-validated error.
Analogously, several vectors with maximal coefficients can be specified
in <code>max.coef.list</code>. In the next example we do not know if it
is better to include variables b4 before those of b3.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>cvm_pl1 <span class="ot">&lt;-</span> <span class="fu">cvm_prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>                             <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, <span class="at">standardize =</span> <span class="cn">FALSE</span>, </span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>                             <span class="at">block1.penalization =</span> <span class="cn">FALSE</span>, </span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>                             <span class="at">blocks.list =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>                                                <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>)), </span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a>                             <span class="at">max.coef.list =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>), <span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>, <span class="cn">Inf</span>)))</span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a>cvm_pl1<span class="sc">$</span>best.blocks</span></code></pre></div>
<pre><code>## [1] &quot;bp1 = 1:4&quot;     &quot;bp2 = 5:9&quot;     &quot;bp3 = 29:1028&quot; &quot;bp4 = 10:28&quot;</code></pre>
<p>We see that the order of blocks which leads to the best result is
specified first in <code>blocks.list</code>. It might be inconvenient to
define a lot of lists in the block argument. Note that it is not the
general idea of prioritylasso to be applied to any order of blocks and
fish for the best.</p>
</div>
<div id="validation" class="section level2">
<h2>Validation</h2>
<p>Now we give a detailed example with a validation.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>pl_fit1 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>                         <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a>                         <span class="at">block1.penalization =</span> <span class="cn">FALSE</span>, <span class="at">max.coef =</span> <span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>), </span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a>                         <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The idea before simulating the data was to include it in the order as
it was defined in the data matrix. The first block, a block with only 4
variables should not be penalized. Because of prior knowledge (we have
it because we generated the data, normally a practitioner could have it,
too) we know that all of its variables are relevant. The maximal number
of coefficients for the last block is restricted to 10. We can easily
extract the coefficients and print those which are nonzero.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a>coeff1 <span class="ot">&lt;-</span> pl_fit1<span class="sc">$</span>coefficients</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>coeff1 <span class="ot">&lt;-</span> coeff1[coeff1 <span class="sc">!=</span> <span class="dv">0</span>]</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">round</span>(coeff1, <span class="dv">4</span>))</span></code></pre></div>
<pre><code>##            b1.1    b1.2    b1.3    b1.4    b2.3    b3.2    b3.3    b3.4    b3.7 
##  0.1683  1.0135  0.9081  1.0682  0.3760  0.1123  0.3115 -0.4062  0.8406 -0.2676 
##   b3.11    b4.1    b4.8  b4.106  b4.269  b4.549  b4.623  b4.748  b4.831  b4.977 
##  0.1103  0.0056  0.0615  0.0407  0.0516  0.0056  0.0863  0.0211  0.0012  0.2553</code></pre>
<p>We see that every variable from block 1 is nonzero, but we also have
variables from every other block in our model.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">library</span>(pROC)</span></code></pre></div>
<pre><code>## Warning: Paket &#39;pROC&#39; wurde unter R Version 4.2.3 erstellt</code></pre>
<p>The R package <code>pROC</code> gives a lot possibilities to evaluate
and visualize the results. First we can calculate the score and the ROC
curve in the training set and get an AUC value.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a>pl1_score <span class="ot">&lt;-</span> pl_train[ , <span class="fu">names</span>(coeff1)[<span class="sc">-</span><span class="dv">1</span>], drop<span class="ot">=</span>F] <span class="sc">%*%</span> coeff1[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>pl1_roc <span class="ot">&lt;-</span> <span class="fu">roc</span>(<span class="fu">factor</span>(pl_out_train), pl1_score[,<span class="dv">1</span>])</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">auc</span>(pl1_roc)</span></code></pre></div>
<pre><code>## Area under the curve: 0.8318</code></pre>
<p>Of course we can’t use this as a performance measure because it is
estimated based on the data that was already used for training the model
and is thus biased. A more appropriate value can be obtained with the
test set.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>val1_score <span class="ot">&lt;-</span> pl_val[ , <span class="fu">names</span>(coeff1)[<span class="sc">-</span><span class="dv">1</span>], drop<span class="ot">=</span>F] <span class="sc">%*%</span> coeff1[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a>val1_roc <span class="ot">&lt;-</span> <span class="fu">roc</span>(<span class="fu">factor</span>(pl_out_val), <span class="fu">c</span>(val1_score))</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">auc</span>(val1_roc)</span></code></pre></div>
<pre><code>## Area under the curve: 0.6471</code></pre>
<p>To investigate the influence of the priorities on the prediction
accuracy, we can calculate the model again with another order of blocks
and get the ROC curve in the same way.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>pl_fit2 <span class="ot">&lt;-</span> <span class="fu">prioritylasso</span>(<span class="at">X =</span> pl_train, <span class="at">Y =</span> pl_out_train, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>                         <span class="at">type.measure =</span> <span class="st">&quot;auc&quot;</span>, <span class="at">blocks =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">28</span>, <span class="dv">5</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">29</span><span class="sc">:</span><span class="dv">1028</span>), </span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>                         <span class="at">block1.penalization =</span> <span class="cn">FALSE</span>, <span class="at">max.coef =</span> <span class="fu">c</span>(<span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span>, <span class="dv">10</span>), </span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>                         <span class="at">standardize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a>coeff2 <span class="ot">&lt;-</span> pl_fit2<span class="sc">$</span>coefficients</span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a>coeff2 <span class="ot">&lt;-</span> coeff2[coeff2 <span class="sc">!=</span> <span class="dv">0</span>]</span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a></span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a>val2_score <span class="ot">&lt;-</span> pl_val[ , <span class="fu">names</span>(coeff2)[<span class="sc">-</span><span class="dv">1</span>], drop<span class="ot">=</span>F] <span class="sc">%*%</span> coeff2[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb46-11"><a href="#cb46-11" tabindex="-1"></a>val2_roc <span class="ot">&lt;-</span> <span class="fu">roc</span>(<span class="fu">factor</span>(pl_out_val), <span class="fu">c</span>(val2_score))</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">auc</span>(val2_roc)</span></code></pre></div>
<pre><code>## Area under the curve: 0.6394</code></pre>
<p>We can plot the ROC curves and perform a paired test if the two AUC
values are equal.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">roc.test</span>(val1_roc, val2_roc, <span class="at">paired=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 
##  DeLong&#39;s test for two correlated ROC curves
## 
## data:  val1_roc and val2_roc
## Z = 0.90554, p-value = 0.3652
## alternative hypothesis: true difference in AUC is not equal to 0
## 95 percent confidence interval:
##  -0.008909984  0.024213666
## sample estimates:
## AUC of roc1 AUC of roc2 
##   0.6470588   0.6394070</code></pre>
<p>The test result shows that we cannot reject the hypotheses that the
AUC values are equal.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">plot.roc</span>(val1_roc, <span class="at">grid=</span><span class="fl">0.1</span>)</span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a><span class="fu">plot.roc</span>(val2_roc, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">add=</span><span class="cn">TRUE</span>)</span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;prioritylasso Score 1&quot;</span>, <span class="st">&quot;prioritylasso Score 2&quot;</span>),</span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="at">lwd=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAtAAAAJACAMAAACT76HtAAAAb1BMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6Ojo6OpA6kJA6kNtmAABmADpmkJBmtv+QOgCQZgCQkGaQtpCQ27aQ2/+pqam2ZgC225C2/7a2///bkDrb///d3d3/AAD/tmb/25D//7b//9v///89x43vAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2diXbbSJJFYY2k6p4R3VJPt8FpwyJF4f+/cbAQxJpALhG5RL57DksyDUcQ5K3EI7YsagAEUYR+AQBQAqGBKCA0EAWEBqKA0EAUEBqIAkIDUUBoIAoIDUQBoYEoIDQQBYQGooDQQBQQGogCQgNRQGggCggNRAGhgSggNBAFhAaigNBAFBAaiAJCA1FAaCAKCA1EAaGBKCA0EAWEBqKA0EAUEBqIAkIDUUBoIAoIDUQBoYEoIDQQBYQGooDQQBQQGogCQgNREAtdcPHJsCRLUbSfU2r/231CCU1bbuSTYUmWomg/o9T+p/uIExokCZXPEBrEAJnP8oROd6ObcXs6nyF0sKJo/4DQZ3lCg+Sg9JlL6O+Pfh/K02+SckAwpD4zCV0Vb/0v1+EXp3ImpLnRzbg9rc88Qn9/PDSunv84lzMiwY806/bEPvMIfTu9D79eFaEDkQO0UPssb4QGKUHuM1uGvg/RyNBov7Mkvc9cezlup34vh2J8htBoX7P4jP3QIBgcPkNoEAoWn9mFrnzv5Uhro5tN++Fs5cudTyaffY/Q4wnezboHfnzqL/vJ0J+jpsk6cTyU67T0+XIp+V6HT6G5yoG4GS4kuVzuTzCNz8jQwAtLodl8lie09iYnsxAbtv1C6NKkvRkQOlTRrNrPhS6N2pvBdC7HeBEuzuUAC6H58kbNNUJ/fyhPhLYpB1JnKjSrz3wn+L9QljMg1o1u3u0nQpem7c3gytDX4n337yF0Vu1HoUvj9maI+1IIYmQ8osLeiXzBIOVA3FwgtC2xbnTzbv/Y/2zR3gwIHapoVu0f+58t2pshTmgQI/z7nwcgNHDncownn+UJHetGV3J7DZ8vC58ROXTJ0qjA7R/7Lnb2Qy/GZwgN4uV4Z5z+jfVdgdDAGQiNyCGq/WHkKFdCI3LokqVRgdsfCV2uR2gIDeLlIHKUiBwgKfaFbvdvQGhrstzmB26/Gzm6/XWIHNZkaVTg9ntC9/ufITRIiMuFboJXVyA0cAZCI3KIaq+MHOrzkRA5dMnSqMDtVULvnF8HoUG8KHbb+TlfdAGEBvY87im69ZdBfJYndJbb/EDti6XQ0yX3fUbk0CUrowK3X90kd7LkwfgMoUF8rIQeCZM3aggNHFALHcxneUJntc0P3F4ZOY59RuTQJSujArdXCa0xPkNoEB+KyBEub9QQGjiwLXRQn+UJndU2P3D7zcih5zMihy5ZGeW7/c59kYYlNcdnCA3Cs3ejrzth80YdTuiwM57iYfVoxJ39uYkci2XK4K/xM5DQiBwJtl8eQFmdtq8/Phu0N0Nc5BBtVOD2R0KXLO3NECc04OMh9PaFVcHzcwuEBtoMQm9fKRiFz/KEFr3ND9x+FHpjqZKrvRkQOlTRBNvvCV2ytTdDnNCAjx2h48gbNYQGGiwPoKyFjsZneUKL3uaHab86Iqjc/4zIQY9Eo4ILvXhitf+Ztb0Z4oQG5BwJHU/eqCE0OOZA6Kh8lie0xG1+1JFj6jMiBz0SjfLdXnl26MYRwtn4DKFBhChPd9445B1X3qghNFijnDIl5v3PA+KEFrHND9t+KfRj0cPznxE56JFgVOD22kKvxmcIDQKimj1iZ4aJ2b+PL2/UEDpjlNZqCh2lz/KElrDN99NeOZGPMnLM2PIZkYOehIwK3N5N6M3xGUKDcGgLvUWceaOG0Blx8BVvREPoaH2WJ3RC23y/RY/2WYwcRw6Vz4gc9MRrVOD2C39t9kMPKMdnCA28oT878VHkiDdv1BA6H8iEjtpneULHu80P1X48XqLNXtE9nxE56InQqLDtLXzeE3p3fIbQgJ2dyQQtiDtv1GxCn4vi5fujGRfeScoBe0iFjt5nJqGr5z+N0y91fTu9EZQzIb5tfuD2CqGt2h/5LDVyfH80Gl9//Kp7tV3LGRGfUZ7bb5w8t5WMbdofjs9Shb6dmqRxffpdD/91KwcMUPvsHDnizxu1wBE6d1a7m2m+DNaJ+MydoTu1ncuZEH6bH7a9rtDG7XV8lho5Qu7lCG5U4PZcQmuNz3KF9l0uY5Tn0JFl55Y08kYNoZNHeVIo3ZfBOiGf2YWufO/lCL7N99x+HjEmSx6IbNRe1+f8IsdjGAk922jz+NRf9pOhP1HN+Wyuk3W6XMheaxnyfTJ/+BSaq1y+OF0ZqEc6eaNGhk6ejb0alF8G68R85hK622XXoEjQyNBk7ZcZWnkSqG37MvTam8F0YKW4H0+5Fjiwwtt+LTRt+zL42pvBd+i7B4e+mVlGDrro3JFW3qg5T07qwclJzPAKnZzP8kZoRA7C9qX+omxLGsKVoe9DNDI0d/v1kUG69qX+onxLGsK0l+N26vdyKMZnRA4yxgtSiPfWpZg3auyHTh7aa2CnJOmzPKFzjRy90ITtR58ROfyXG4HQNO0n4zOE9l8uP5ani+a+/3kAQqfJ6vzn3Pc/D4gTOpPIobjSiqj93GdEDv/lRiC0e/vF+Ayh/ZfLDr67FaScN2oInQymtwy1J2mf5QktNHJo3gOXoP3aZ0QO/+VGxAo9f0px5ZV7+43xGUL7LyceTaGdSTtv1BA6FTwJnbzP8oRG5HBov+0zIof/ciMQ2r69YnyG0P7LicdH5Eg/b9QQOhU8CC3CZ3lCI3JYtlf7jMjhv9yINKEfE0tozTxv335nfIbQ/suJpTAT2hoZeaOG0LHDd8ngDDE+yxNaWOTohN7I0LTt931G5PBfbgRCm7c/GJ8htP9yYvEROeTkjRpCx44HoUX5LE9oRA7D9sc+I3L4LzcCoc3aa4zPENp/ObFwRw5ZeaOG0LHDLLQ4n+UJLS1ymF0Da9hez2dEDv/lRoQJbXhRt1l7zfEZQvsvJxbsfzYEQscN9j8bEkro0NPnfiYyNXIrNM86lcHWifsRRGhk6P0lZ9mZo73B+Ox/7R1A5IgShjt8zZGZN2oIHSnMpz/L9Vme0DIix1xo8vZG8x0jcgQoNwKhjymNakLoAOXE8LiEkLGH3LxRQ+jIKDwILdpneUKnHTm2z0UibW883zEiR4ByIxB6n2F8jnHtKRAndNqwX3ElO2/UEDoyDM8WNUa8z/KETjtybB8gJGtvNd8xIkeAciOpC83Zfjo+x7j2FIgTOm2w/9kVCB0V2P/sijihE40cuwdUSNpbz3eMyBGg3EiaQu8fIaRovxyfY1p7SsQJnSbY/0wFhI4C3E+GCnFCpxo5+p9MkcNpvmNEjgDlRiD0mq3xOaa1p0Sc0GnCGjnyyRs1hI4ETqGz8lme0IgcC5znO0bkCFBuBELPUY3PMa09JeKEThO2yJFX3qghdCCWs2hyCZ2dz/KETiJyrH3miRwk8x0jcgQoN5KI0F7Oh94bnyF0oL4iUc7UTRo58ssbNYQOgxehs/SZS+jz0+/667UofvwiKWdA3JFjEpm5IwfZfMeIHHef/2pkvp3eCcqZELXQhT+hj8ZnCG1Q9HZ6a6R+aX+tnv84l5ODMmoMkEWOPPNGzSb0e/398db+em3GatdycvAmdLY+c0WOZnSuwozQkUeOgyWJIgfpfMeIHO0Q/fS7G6Kvqm+FEJpRaJ3xGUKb1b32331eiMoJwU/kyDdv1NgP7RcvQmftszyhs48c5PMdI3JMqXzv5chdaN3xGULTtBsIPdtomMfhel8ujj3K4OsY/uFTaK5yicCeofPOzy3I0D6LckeOMuq1t2xvBpPQ3x99slAkaAjNInQZ99pbtjeDR+iqeOt/uQ6/OJUTA2/kQN6omYS+n8fRgpOTWi762DeBzy1sJyfd8X5yUowbXQOf7SNHSfBCXRdNO3LcTqoj2yFH6Bg/0oemhxnavn1pWBRCbzxXNV/7NjNyVdyHaGTolk2haUHeuOMaORRON6N3h2J8htDEwOcB9wxd7e2dc+9rSlQb3WUy5oocD5+jWnui9mY4Ct2eJdpenaIciZ37mhLTR7r6qsck9Dg+x7T2VO3NcBG6zRW9yap9GQR9U2a1z4InciBvTHDay6G8SQFh35TxIzR8nuIi9M/eZ/Ph2aSvKTFtdHeFpmo/8zmmtadqbwaB0KpTnmn6mhLTR+pD6Pn4HNPaU7U3w1ro83jTFMWuZpq+KeMhciBvLCAYoXn7pgy/0PB5Cc6HZizKHjlWPse09lTtzYDQjEW5hV6PzzGtPVV7M2yFvp3ehsPbFscJETkoQN7YQNwIHROTI4TzmSdIgM9biBM6po3uIPRqKhWK9ps+x7T2VO3NcDpSqLzRF2FfU2L6SEeh6dtvj88xrT1VezNcRuh2V7Tt0e+sIgfDEW/kDQUEZ9up78hI0Tdl+ISGzyrcM/T3R1R7OWLa6F4uO98FXdorfY5p7anamyFuhI7pI+USWj0+x7T2VO3NcBC6u5kMMvQO5DN39yBv7CBuL0dM8AgNn/fAfmjGortC27bf9TmmtadqbwaE5iiqcx8ky/b743MUa0/c3gycy8EAxY29FCBvHCBuhI4Bpi+DNXw+BtcUMhTVEtqm/aHPMaw9dXszcE0hQ1EuoY/H5xjWnrq9GbimkBK+7NyCvKEBrikkhPHLYA2f9RD3pTDkRrcRma+9ls+IHBCacFFOofXGZwiN/dCE8O2uQ97QRdwIHQLe7NwCn3UJJXTo2Uabx6f+sp+7fz+dIYWq5vxRMqwTx8Nknfge1kI3saO93bnxvaEV5WjQXiG6FDkdmDnal/rjc4C1Z29vhtM1hc9/vl5f6nNUJ/gHgDE6tyBumOB0PvR7fS3eIzv0HQBeoeGzEY5CnxuZ8z30vfFdkLx9yVHUaMnQ7c1wihwvt9PzH7sLVyQIvbVzg7o923SaEHr91O1U/Pj1/WF1IZaEyMF4240B5A1TsB/aHn6h4bMx4oT2Gjn6n3Z3rNNYcvAZkUMfF6G7+xjEduhbkNCs02lC6PVTZxuTTftGDHPkQN6wwWm3nc2Z/aZ9I4ZXaPhsheN+aP6+pkiJHFOfETn0cboVWIx3ThIiNPt0mhB6/VR73Ju9b8Tg9s8R4hQ5Mj/BH7d/jhDsh7ZflC1yLH1G5NAHQtsvyiW0l+k0IfTGc03oeP5zttp5h8ihBHnDBacvhT9+Ve3ZdrneaIZHaPjshNNuu7e6ETrf86FZIseWz4gc+jgeWGmFjuuKlcSF9jadJoRePTOM0Gebq2QROTZB3nDFPUNXVodXIPQW8NkZ170ctvNgIXJsLKnyGZFDH+yHtlj0cS2hy5T06yW9TqcJoQP1jZDLjtAOIG9QYC/01+tbN5NsZJHDA9j/HDHWQn+9vvfn23W/8PU1xcNGd09o6/Z7PiNy6GM/JUV7MnR3EVZctwJLVGjv02lC6PkfuwPe/Sn+cR1YYWRIzMTZuQV5gwp7od+HqwpzEbrgExo+k+EkdB+fq6iOFPJtdB/+7t2c0ar9kc+IHPo4Zeiq3cOxc2nhzvdFCD0uGWQ6TQi9+PP1x6/2Vo2bd+d4XJylvj4rychx/4X49rnIG5TY74e+dvfu/3rdChzXfjLOECM0HwXPVCrwmRSmI4X94C0rclx0hDZur+MzIoc+bIe+z02+liY0Q/tg02lCaNPKVfEmK3JwzDyBvEEN48lJX6//BaH3gc/kcJ5t9/2hPvcfkeNT32dEDn3EnT6akNDa4zOE1odbaNUl4YgcyBsseB6hH8dbQk+fa/G4XEjraU93jIfhw6fQXOVGtFcocOQoTdsHWzJ0ezOQobUXJRW6TMgoCK0xn1CEGbpQwXDEG/mZCx6hq+J+v7trobjxXXxCH/tMJzR8ZoNF6PaeSndUJ0vHFzmUp+s3IhNvSg3nOw69zUfkmEwnpLqcJWehS+0lDYryLRm6vRniRmhr9oQmBXmDE64MfR+iU8jQj7A8PHFZQtoOPrPCdj5074jycsN4IkexFHrlM2nksJjvOPQ2H5HDf7kRC6EXz2+MyHQf1Dg+p2MUhPZfzgEdoclA3uAGQvsUGj6zI05o+8ix8yWQaFNqOd9x6G0+Iof/ciPWQu/t1aD5oObjczpGQWj/5RwYhWZuhLzhAwjtS2j47AVxQjtEDsKiaxzmOw69zUfk8F9uxFxojtvHrFiPz+kYBaH9l3OA6Qj3HOQNX0Bo9i+DNXz2iDihjbd6OkI7bkod5zsOvc1H5PBfbiRCobfH53SMgtD+yzmA/c+igNDY/ywKcULHFjkI5jsOvc1H5PBfbiQyodXjczpGQWj/5RzA/mdRQGiczy8KcULrb8se95GhLDpbkmi+49DbfEQO/+VGtN+pgl3o/fE5HaMgtP9yNq+A+7RR5I0QQGguoeFzEMQJbRA5+p9MkYNwvuPQ23xEDv/lRiIR+nh8TscoCO2/nM0r4IwcyBuhgNCYT1MU4oSOIXIQz3ccepuPyOG/3EgEQuuNz+kYBaH9l7N5BVyRA3kjJBAa9zMXhTihQ0cOhvmOQ2/zETk0CD3b6OfnMJvt5XK87Kd23VK7v35Ng8dn2PeUZZ2MH0GEFho5kDdCIy5y6L8CBqHhc3DECa2/yaHP0KVB+4RCbOD2ZkBoOqFLk/YJGQWh/ZezeQXUkQN5IwYgNJXQ8DkKxAmtHzkM7jqqUbQ0bJ/QNh+Rw3+5Ee13yuQ2usdFh/FZolEQ2n85C7D/WSQQmgL4HA3ihDaIHGRFJz5L3OYjcvgvNxJA6On4LNEoCO2/nAXY/ywSCO0KfI4KcUL7jhwLnyVu8xE5/Jcb8Sz0cnyWaBSE9l/OApLIgbwRGxDaBfgcHeKE9hk5NnyWuM1H5PBfbsSj0Fvjs0SjILT/chY4Rw7kjRiB0LbA5ygRJ7SvyKHwWeI2H5HDf7kRT0KrxmeJRkFo/+UscIocyBuxAqFtgM/RIk5oH5Fjx2eJ23xEDv/lRjwIvTc+SzQKQvsvZ4F15EDeiBkIbQp8jhpxQnNHjgOfJW7zETnquiqK4r375ek3QTkDmIU+Gp8lGgWh6+rHr/p2eqkDCK2NVeRA3ogdFqG/P966/z7/ESY0fI4eFqFvpy5u1OfnP6Iih4bPErf5iBz9CN1wfpEktM74LNEoCP0IGrdTISdyIG+kANtejj50fH9EJ3Rxx1Ro+JwE2e2HLmyE/tT2WeI2H5HDf7mRY6Hvv5gJrTs+SzQKQk+IbredldDIG8ngeYR+bPCDzTTaCN3/rjOD7PAog8+Oiof+w6fQXOVGlCv0+H/p/meDEbrUfpdEbvMDtzcjmwxd2Atdhv5I825vBpPQ3x+9PIoEHSBDF8uO2kIjP6cE04GV4n6o8Dr84lSOAmuh4XNS8B76btR+/uNczgh15Fg8oSl0uVdUv73Lkpm3N4P15KSGayznckyys/78hMP4nLdREDrkCK1inAnZRGjkjdTgytD3ITq+DI3jKbJh2stxO/V7ORTjc8DIMQp9vNV7+Jz3Nh+Rw3+5EUKhx/E5b6MgtP9yGg2NIwfyRorkI7TRzo0aPieKOKFV27L1zo39rd7M57y3+Ygc/suNqIXWXbJjPj7nbRSE9l/uGLMrrpA3UgVCbwGfk0Wc0BSRY+Vz3tt8RA7/5UYIhF6Pz3kbBaH9lzsG+5/zAEIvgc9JI05o18ix6XPe23xEDv/lRhyF3h6f8zYKQvsvd4zJ+fwgXSD0FPicPOKEdokcSp/z3uYjcvgvN+IgtHp8ztsoCO2/3DHHkQN5QwIQegA+i0Cc0LaRY9fnvLf5iBz+y41YCr0/PudtFIT2X+6Y/ciBvCEF+ULrzEABn8UgTujltkw9pcq45KHPeW/zETn8lxtZC93/3BH6eHzO2ygI7b/cTqPD+3Egb0gCQmfucyEN7RUnfh9py42YRg4tnwVv873vbmImd6H1xmcInQzihF412o0cmeeNGkJ77+vcaE9o+Ayhffc1xSRyaPuMyJEMOQutPz5D6GQIJbS3qXKHqZAboRd/V4aewjeKB4T23Ne5kWqERn7ugNCe+5qiGzlM5jtG5EiHXIUuozEqcHsI7bmvc6NNoZE3BiC0577OjbaEhs8PILTnvqboRA7T+Y4ROdIhR6HL7SUNivpdEkIbIE7oVaOV0MgbUyC0577OjZZCw+cZENpzX1NWkWMxP6HNfMeIHMfcTu9bz1zf1Qvs1CqK4um3du+vv43Lihd6MeFmqV7SoKjfJZMQepuJxfpCX4t2yfOPX9pNngQLvQT7n/eJUOjzS/vf748XvcWvs9E8L6Hh8wr7D+Lrr3+9FsVbo+rP/y2e/tP52sg1fearWeD5n62Y1Usn9Ln5+3a59i/aX4afwz9smZlcNc+/LAr/7p69L938zdtVstDryDH+bjvfcU6RQ/8C1EbG9y4e3E7Pf+6JufvTy/hM+2h1+/54b39tx96qiRJfr/3Cw8/HP+y4Fg+j24Vvp7d54XuJt8cryVXocn9Jg6J+l/QptMEV1b1S1fOf1rhO3e+P9pnrj1+PZ4bH11/Nc++3n7/qTubBwOHn4x8Ole8poqsz/ft74bd6ZrFooZdg//M+DpHjtcsKrWXvvb79M81/H8/0OeOl177L0NeiG2x7BYefj3/4KP79UQwj+fTv++1AJ/5k6TyFhs+buGTo1qu10O1QPBO68e381v3aZN+n/2sXanztsvP95+MfTuufn373LepF4et9e5GH0KrI4TLfcUaRwwCV0KsR+vbz3z/bKNxH5sfIOuyXa37ORuhhgevw/PYIPSFHoTfG50iMCtzeJXK0UfZ8DxOLDD0V+vvj7/13xE686zCyDgNy83OWoYe9HM3SGxn6fWMHoGihB+YHVJA3VLgI3QjY734YLBt3RjyE7r44Fv13xHagvZ2aP3TqNv8Zfi73crS7Mq7tP2r3Z7SCzwp3z04PvOQg9GUmNHxW4iL0P167IDsZNofdxY9nzq2cXTi5Z+gfv86dwO1v9ePndD/0/dB37+tyP3Q/OFfzQ+Oihb5vSjWOp0SyzQ/c3nUvh9aS//3HuospOQitGJ8jMSpwex9CV2/Hy1AhTug76uMpYAa/0F+vz/4GaHlC3/dR4niKJjgf2nNf07pLoSnmO0bkSAd5Qvc/1cdTHkRiVOD2ENpzX9O6c6GRN46A0J77mtadCQ2fD4HQnvua1p0KTTXfMSJHOogTenKA8GB8jsSowO0htBbdaYF7l+5yvY+TI97IGzokf9V3p9p45IZH6GrocC0UB4n4hB5+g89aRHiRrNFV398fzXLVeNEWi9D92X4dleIoEbvQlPMdI3LYwn/Vd3+8snqM5yxCT177VbHl4BZaY3yOxKjA7ZO/6rtjPOdf6AiNvKHL8oO47DFbMp6rvrvLtRTro73iu1TDVQmBMjR81mbxQez6vBQ6lqu+p5ox7eXovqc2KM+zYhWaer5jRI4tIrrqexzSxe2H1tj/PBCJUYHbC7jqexYDBAqNvGFC+ld9V7NYyy105X0vB3w2IvmrvqtiNq57HqEfN5Ximhm1jRyay37q1/1keK0cNU3WqXadSTaOq77nezvkRY4yuRAbuH3qV31X/QDJux/af7kB5A1TcNW3FoFOToLPxuCqbx0CnZzU+pzcNj9we1z1rUGgQ9/d+JycUYHb43xoDcKcnIS8YQOE1iDICA2frYDQOgQ4OWnwObltfuD2EFoL7ycnPcbn5IwK3B5Ce+6rB/KGLRDac18t4LM1ENpzXx2mPie3zQ/cHkJ77qvBbHxOzqjA7SG0577HIG+4AKE99z0EPjsBoT33PWLpc3Lb/MDtITRRXyLK5ROf2v9Uf0mWoum0T4tAQidVM+8XKnTtU1mr8O9UwJrpFA3/QlNZq/DvVMCa6RQN/0JTWavw71TAmukUDf9CU1mr8O9UwJrpFA3/QlNZq/DvVMCa6RQN/0JTWavw71TAmukUDf9CU1mr8O9UwJrpFA3/QlNZq/DvVMCa6RQN/0KlHSIFmQOhgSggNBAFhAaigNBAFBAaiAJCA1FAaCAKCA1EAaGBKCA0EAWEBqKA0EAUpEJ//W24r383uxDJnDJjzXYOsPlko5bM6rTzh+nOkqNbs52l/WVvaZuidTcbJm1Rqo9p9kKpPia7l0op9DBnedu9eSEVxYc61uxnciR4q2Z1qvv0kJQ1r90Eq+4rv1zhq/q+3HZFqT6m+doTfUyWL5VQ6Os43Vs/k5Jq/mS7mv2EGGfnd39Wp5+v17noomY3zarzyi9X+HYiEHpWlOhj2lh794/J9qXSCX0t3hZTCbn/fzqtSfU/yawOkdCzmv3s7+4sV7h6/qe70BvvovPHNF97qrHM8qWSZuiF0GeCtRpr9p6opt/SZ16HJnLMal6f/nOiSKaLFW7+SJChN95F549pVpPqY7J9qYxCK2cWsqvZ/99JMOrP65B8hZnVrNqU1A/9dEW7DTCB0Ot30f1jmtWk+phsXyqf0FeSL/rcQrf/z3+9kn6k1Q+aMWr+QtuJ9DiEJviYfAmt9VLZhKYZn7kjB1Him9Xsq+lPma1VtPsDQ+Sg+Jg8RQ69l8oldEXkc837pZBoPJnX7IV2/mo4K1rd7yrr/H/J/F0k+ZjovxS2a/tm91KZhK4oDlbMa7LstuvfNOfxZFazn0HafYxarzDBCD0vSvMx+dhtp/1SeYR2T6XrmjwHVmgy9OJgTePdZBppmqItFEcKp0WpPiYPB1a0Xyq90O33+/sGkmK1HjW7DRHJG3Wvc98HTXPwd1bzSnTcf1a0Jjr0PSlK9jHNXijVx2T3UnFyEhAFhAaigNBAFBAaiAJCA1FAaCAKCA1EAaGBKCA0EAWEBqKA0EAUEBqIAkIDUUBoIAoIDUQBoYEoIDQQBYTm4nYqxvuYqRZ57y7L+J/T+/y5K9UVmfkBoZnob8d0Prxm6HZaX6x1O0FoWyA0E/3Vysc3UNqSF0LbA6F5mJj89de/Xu8XzVbF5JeXVtz2VtJP/+kEnj33/M/uelOSm09lBYRmYrxvVePne59A2puEddfjt780WaMdiYfH8rlrd4M8jNSmQGguurG3vylYd7/o5z99XG5MHXLzVAJtVCMAAAESSURBVOit56huzJsTEJqR74+iG5O7Oyn9+NXfNKX543Dfu5m8i+faEF5R3IcjMyA0L+en3/ebDjZCD/enG0berdH4IXQzkp/J7j+VDxCah2HAbTwehR724WmN0Lef//6JxGEMhOZh2MvRjLN9hj63Gfr+HU8nQzcV/o7EYQ6EZqKfsqq9p/HX6zDtRXcr9PZYS/tLo/xqL8fjue5rJNUdibMCQnPRHfpuDf56/cfr/cbO1XA0/LHPeRR68lx9bv9vwD4OGyA0O7b38v/6byQOcyA0O7ZCV0gcFkBoduyE/nrFV0IbIDQQBYQGooDQQBQQGogCQgNRQGggCggNRAGhgSggNBAFhAaigNBAFBAaiOL/AUqz2fMJOmLvAAAAAElFTkSuQmCC" /><!-- --></p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
